src/edgepi/peripherals/gpio.py

```diff
@@ -45,6 +44,66 @@ def read_state(self):
         """
         return self.gpio.read()
 
+    def open_read_state(self, pin_num:int, pin_dir:str, pin_bias:str) -> bool:
+        """
+        To minimize issues with the lock, we open & read in a single function call
+        """
+        try:
+            # pylint: disable=consider-using-with
+            GpioDevice.lock_gpio.acquire()
+            gpio   = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+            result = gpio.read()
+
+        finally:
+            try:
+                gpio.close()
+            except Exception as exc:
+                raise OSError(f"Failed to close {self.gpio_fd}") from exc
+            finally:
+                GpioDevice.lock_gpio.release()
+
+        return result
+
+    def open_read_state_batch(
+        self,
+        pin_num_list: list[int],
+        pin_dir: str,
+        pin_bias: str,
+    ) -> list[bool]:
+        """
+        Batch several gpio reads into a single lock / unlock. We can also take advantage of
+        the fact we're accessing the gpio only from different pins, and so we can open & close
+        just a single time.
+        """
+        results = []
+        try:
+            # pylint: disable=consider-using-with
+            GpioDevice.lock_gpio.acquire()
+            gpio = None
+            try:
+                for pin_num in pin_num_list:
+                    if gpio is None:
+                        gpio = GPIO(self.gpio_fd, pin_num, pin_dir, bias=pin_bias)
+                    else:
+                        # NOTE: we'll need to be careful when we update periphery, since we depend
+                        # on private functionality
+                        gpio._line = pin_num # pylint: disable=protected-access
+                        # pylint: disable=protected-access
+                        gpio._reopen(
+                            pin_dir, edge="none", bias=pin_bias, drive="default", inverted=False
+                        )
```

Comment:
It's a performance gain of roughly 40%.

If you look at the following image, you'll notice that `posix.open`, `_reopen`, `read`, and `close` all take roughly the same amount of time (read is the shortest, but luckily required). When we batch the reads, the main performance improvement comes from the fact that we don't call `posix.open` and `close` 8 times. Instead, they're only called once.

![Screenshot 2024-08-13 111629](https://github.com/user-attachments/assets/e004c6e7-9468-4eb9-8358-02b746766ffd)

The number 40% is because that's roughly how much of the hot loop `close` takes up.

Reply:
please add this as a comment: 

> If you look at the following image, you'll notice that posix.open, _reopen, read, and close all take roughly the same amount of time (read is the shortest, but luckily required). When we batch the reads, the main performance improvement comes from the fact that we don't call posix.open and close 8 times. Instead, they're only called once.

Reply:
okay! Good idea

https://github.com/EdgePi-Cloud/edgepi-python-sdk/pull/423/commits/2cb3644b619ad62316c875e225f6ab37a1b04afc